components/vocabularies/hpo/api/src/main/java/org/phenotips/vocabulary/internal/solr/HumanPhenotypeOntology.java

```diff
@@ -103,6 +117,51 @@ protected int getSolrDocsPerBatch()
         return result;
     }
 
+    /**
+     * Returns the same JSON as {@link VocabularyTerm#toJson()}, with the exception that the categories are ordered.
+     * @param term whose JSON should be returned
+     * @return JSON of the `term` with ordered categories
+     */
+    public JSON termToOrderedJson(VocabularyTerm term)
+    {
+        JSON rawJson = term.toJson();
+        JSONObject json;
+        if (!rawJson.isArray()) {
+            json = (JSONObject) rawJson;
+        } else {
+            return rawJson;
+        }
+
+        Set<VocabularyTerm> categoriesSet = term.getAncestorsAndSelf();
+        List<VocabularyTerm> sortedCategories = this.sortTermsByLevel(categoriesSet);
+        List<String> sortedCategoriesIds = new LinkedList<>();
+        for (VocabularyTerm category : sortedCategories) {
+            sortedCategoriesIds.add(category.getId());
+        }
+        json.put("term_category", sortedCategoriesIds);
+        return json;
+    }
+
+    private List<VocabularyTerm> sortTermsByLevel(Collection<VocabularyTerm> terms)
+    {
+        List<VocabularyTerm> sortedTerms = new LinkedList<>();
+        Map<Long, VocabularyTerm> levelMap = new HashMap<>();
+        for (VocabularyTerm term : terms) {
+            levelMap.put(term.getDistanceTo(this.getRootTerm()), term);
```

Comment:
Maybe compromise. You could make the level map (but swap the keys and values) then sort your list based on each terms value in the map. Then your level map acts as a cache for the ``getDistanceTo`` method.